tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -55,83 +62,100 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
+      """
+      orthogonalize `vector`  to all rows of `krylov_vectors`.
+      Args:
+        vector: Initial vector.
+        krylov_vectors: Matrix of krylov vectors, each row is treated as a
+          vector.
+        iterations: Number of iterations.
+      Returns:
+        jax.numpy.array: The orthogonalized vector.
+      """
+      vec = vector
+      for _ in range(iterations):
+        ov = jax.numpy.dot(
+            krylov_vectors.conj(), vec, precision=precision)
+        vec = vec - jax.numpy.dot(ov, krylov_vectors)
+      return vec
 
     def body_lanczos(vals):
-      current_vector, krylov_vectors, vector_norms = vals[0:3]
-      diagonal_elements, matvec, args, _ = vals[3:7]
-      threshold, i, maxiteration = vals[7:]
-      norm = jax.numpy.linalg.norm(current_vector)
-      normalized_vector = current_vector / norm
-      normalized_vector, krylov_vectors = jax.lax.cond(
-          reortho, True,
-          lambda x: jax.lax.fori_loop(0, i, body_modified_gram_schmidt,
-                                      [normalized_vector, krylov_vectors]),
-          False, lambda x: [normalized_vector, krylov_vectors])
-      Av = matvec(normalized_vector, *args)
-
-      diag_element = jax.numpy.vdot(normalized_vector, Av)
-
-      res = jax.numpy.reshape(
-          jax.numpy.ravel(Av) -
-          jax.numpy.ravel(normalized_vector) * diag_element -
-          krylov_vectors[i - 1] * norm, Av.shape)
-      krylov_vectors = jax.ops.index_update(krylov_vectors, jax.ops.index[i, :],
-                                            jax.numpy.ravel(normalized_vector))
-
-      vector_norms = jax.ops.index_update(vector_norms, jax.ops.index[i - 1],
-                                          norm)
-      diagonal_elements = jax.ops.index_update(diagonal_elements,
-                                               jax.ops.index[i - 1],
-                                               diag_element)
-
-      return [
-          res, krylov_vectors, vector_norms, diagonal_elements, matvec, args,
-          norm, threshold, i + 1, maxiteration
-      ]
+      krylov_vectors, alphas, betas, i = vals
+      previous_vector = krylov_vectors[i, :]
+      previous_vector = jax.lax.cond(
+          reortho, lambda x: iterative_classical_gram_schmidt(
+              previous_vector,
+              (i > jax.numpy.arange(ncv + 2))[:, None] * krylov_vectors),
+          lambda x: previous_vector, None)
+
+      beta = jax.numpy.linalg.norm(previous_vector)
+      normalized_vector = previous_vector / beta
+      Av = matvec(jax.lax.reshape(normalized_vector, shape), *arguments)
+      alpha = jax.numpy.vdot(normalized_vector, Av, precision=precision)
+      alphas = alphas.at[i - 1].set(alpha)
+      betas = betas.at[i].set(
+          beta)
+      next_vector = jax.lax.cond(
+          reortho, lambda x: Av, lambda x: jax.numpy.reshape(
+              jax.numpy.ravel(Av) - jax.numpy.ravel(normalized_vector) * alpha -
```

Comment:
Av.ravel() and so on might make this a bit easier to read. I'd also recommend putting the branch functions on separate lines, but it's a matter of taste.